### ISSUE TITLE
wildcard path filenames

### ISSUE THREAD

**Issue by FernandoMiguel** [Neilpang/acme.sh#1408]
if an user does
`$ acme.sh --issue --dns dns_cf -d *.test.imperialus.house --dnssleep 60`

the cert conf path becomes
`/Users/fernando/.acme.sh/*.test.imperialus.house`

dealing with the caracter * in paths is hard, and not safe for many operating systems

can we replace *. by something else?

**Comment by Neilpang** [comment]
>> dealing with the caracter * in paths is hard

example ?

**Comment by FernandoMiguel** [comment]
@neilpang file explorers, some shell scripts, etc.

It's a special caracter, used for shell expansion

**Comment by Neilpang** [comment]
yes,  but we never recommended the user to use the file in our .acme.sh home folder.   We handle the filenames internally.    The user should specify the filenames by `--installcert` command.

**Comment by itssimple** [comment]
In fact the script recommends you NOT to use the files in the .acme.sh-folder, since the path could be moved/changed at any time. :) So as @Neilpang says, make use of the --installcert script, and provide file names that does not contain the asterisk. :)

**Comment by dwatrous** [comment]
I found a use case where this breaks. I created a deploy script for kubernetes and I need to base64 encode the fullchain.cer and the key. This worked until I ended up with a path that encompassed a top path. Here's an example. Consider the following wildcard domains that exist

`*.myapi-cicd.eng.houston.k8s.com`
`*.eng.houston.k8s.com`

For the first domain, I get

```
[Thu Jul 12 23:02:18 UTC 2018] _cfullchain value is /acme.sh/*.myapi.eng.houston.myk8s.com/fullchain.cer
[Thu Jul 12 23:02:18 UTC 2018] _ckey value is /acme.sh/*.myapi.eng.houston.myk8s.com/*.myapi.eng.houston.myk8s.com.key
```

when I create the base64 encoding for kubernetes, like this

```
_tlscrt="$(cat $_cfullchain|base64 -w 0)"
_tlskey="$(cat $_ckey|base64 -w 0)"
```

I get just that one fullchain.cer and key. However, for the next domain I get

```
[Thu Jul 12 23:02:18 UTC 2018] _cfullchain value is /acme.sh/*.eng.houston.myk8s.com/fullchain.cer
[Thu Jul 12 23:02:18 UTC 2018] _ckey value is /acme.sh/*.eng.houston.myk8s.com/*.eng.houston.myk8s.com.key
```

The code to create the base64 encoded crt and key now gets all of the fullchain.cer files and all of the key files currently on my system. The following command shows how this happens

```
~/.acme.sh/deploy # ls -la /acme.sh/*.eng.houston.myk8s.com/fullchain.cer
-rw-r--r-- 1 root root 3896 Jun 18 19:47 '/acme.sh/*.myapi.eng.houston.myk8s.com/fullchain.cer'
-rw-r--r-- 1 root root 3835 Jul 12 22:15 '/acme.sh/*.eng.houston.myk8s.com/fullchain.cer'
~/.acme.sh/deploy # ls -la /acme.sh/*.eng.houston.myk8s.com/*.eng.houston.myk8s.com.key
-rw-r--r-- 1 root root 1679 Jun 18 19:45 '/acme.sh/*.myapi.eng.houston.myk8s.com/*.myapi.eng.houston.myk8s.com.key'
-rw-r--r-- 1 root root 1679 Jul 12 22:13 '/acme.sh/*.eng.houston.myk8s.com/*.eng.houston.myk8s.com.key'
```

I'm now trying to figure out how to make my Linux cat command operate only against one directly rather than interpret the asterisk as a wildcard and matching all folders. 

Can you think of another way that my deploy script should access the fullchain.cer and key for that domain?

**Comment by dwatrous** [comment]
I was able to work around this case with proper quoting in my command

```
  _tlscrt="$(base64 -w 0 "$_cfullchain")"
  _tlskey="$(base64 -w 0 "$_ckey")"
```

I guess it isn't a big deal...

**Comment by scher200** [comment]
@dwatrous would you mind sharing your script, I am looking for a script, just like that.

**Comment by dwatrous** [comment]
@scher200 I didn't take the time to add lots of documentation, but the script should be easy to follow if you're familiar with other deploy scripts. https://github.com/dwatrous/acme.sh/blob/5849cde3098e6f02c8e82d2e420f965a29aa60bd/deploy/kubernetes.sh

**Comment by scher200** [comment]
thank you @dwatrous
It looks well and readable to me, I'll check it out soon!

**Comment by buforded** [comment]
I ran into this today.  I was positive I had done something wrong because asterisks in filenames is so very, very wrong.

Look - I understand you're auto-updating your LE certs with a button on your phone through the facebook app and that is great!  Good for you!  All love!

However, some of us are still using shell scripts in the shell (!) and redirecting and pipelining output and, generally, doing UNIXy things with our tools.  This broke for me in ways I'm not going to waste your time describing.

Please don't output filenames with asterisks in them.

@dwatrous should not have had to write bespoke scripts to work around that brokenness.

Thank you.

**Comment by tpetersvw** [comment]
> I ran into this today. I was positive I had done something wrong because asterisks in filenames is so very, very wrong.
> 
> Look - I understand you're auto-updating your LE certs with a button on your phone through the facebook app and that is great! Good for you! All love!
> 
> However, some of us are still using shell scripts in the shell (!) and redirecting and pipelining output and, generally, doing UNIXy things with our tools. This broke for me in ways I'm not going to waste your time describing.
> 
> Please don't output filenames with asterisks in them.
> 
> @dwatrous should not have had to write bespoke scripts to work around that brokenness.
> 
> Thank you.

I'm dealing with the same issue, can you provide me with any info on how you dealt with the paths/filenames with * in them, that would help me greatly!

**Comment by tpetersvw** [comment]
any reaction from the devs at all would be great seeing that this issue has been open since 2018 (5 years).

**Comment by neilpang** [comment]
The internal cert dir is named after the literal main domain by design; changing the naming now would break every existing installation and every script referencing those paths. The internal dir is not a public interface -- use `--install-cert` to copy the files to any path you like (or a deploy hook), then no `*` appears in your paths. On POSIX filesystems `*` in a filename is valid and only needs quoting.
